# Library importation

In [2]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

# Data Exploration

In [3]:
df = pd.read_csv("spotify_data.csv")

In [4]:
df.head()

acousticness                                   artists  danceability  \
0         0.995                       ['Carl Woitschach']         0.708   
1         0.994  ['Robert Schumann', 'Vladimir Horowitz']         0.379   
2         0.604                   ['Seweryn Goszczyński']         0.749   
3         0.995                      ['Francisco Canaro']         0.781   
4         0.990  ['Frédéric Chopin', 'Vladimir Horowitz']         0.210   

   duration_ms  energy  explicit                      id  instrumentalness  \
0       158648  0.1950         0  6KbQ3uYMLKb5jDxLF7wYDD             0.563   
1       282133  0.0135         0  6KuQTIu1KoTTkLXKrwlLPV             0.901   
2       104300  0.2200         0  6L63VW0PibdM1HDSBoqnoM             0.000   
3       180760  0.1300         0  6M94FkXd15sOAOQYRnWPN8             0.887   
4       687733  0.2040         0  6N6tiFZ9vLTSOIxkj8qKrd             0.908   

   key  liveness  loudness  mode  \
0   10    0.1510   -12.428     1   
1    8    0.0763   -28.454     1   
2    5    0.1190   -19.924     0   
3    1    0.1110   -14.734     0   
4   11    0.0980   -16.829     1   

                                            name  popularity release_date  \
0                    Singende Bataillone 1. Teil           0         1928   
1       Fantasiestücke, Op. 111: Più tosto lento           0         1928   
2                 Chapter 1.18 - Zamek kaniowski           0         1928   
3  Bebamos Juntos - Instrumental (Remasterizado)           0   1928-09-25   
4    Polonaise-Fantaisie in A-Flat Major, Op. 61           1         1928   

   speechiness    tempo  valence  year  
0       0.0506  118.469   0.7790  1928  
1       0.0462   83.972   0.0767  1928  
2       0.9290  107.177   0.8800  1928  
3       0.0926  108.003   0.7200  1928  
4       0.0424   62.149   0.0693  1928

In [8]:
df.shape

(169909, 19)

In [6]:
df.describe()

acousticness   danceability   duration_ms         energy  \
count  169909.000000  169909.000000  1.699090e+05  169909.000000   
mean        0.493214       0.538150  2.314062e+05       0.488593   
std         0.376627       0.175346  1.213219e+05       0.267390   
min         0.000000       0.000000  5.108000e+03       0.000000   
25%         0.094500       0.417000  1.710400e+05       0.263000   
50%         0.492000       0.548000  2.086000e+05       0.481000   
75%         0.888000       0.667000  2.629600e+05       0.710000   
max         0.996000       0.988000  5.403500e+06       1.000000   

            explicit  instrumentalness            key       liveness  \
count  169909.000000     169909.000000  169909.000000  169909.000000   
mean        0.084863          0.161937       5.200519       0.206690   
std         0.278679          0.309329       3.515257       0.176796   
min         0.000000          0.000000       0.000000       0.000000   
25%         0.000000          0.000000       2.000000       0.098400   
50%         0.000000          0.000204       5.000000       0.135000   
75%         0.000000          0.086800       8.000000       0.263000   
max         1.000000          1.000000      11.000000       1.000000   

            loudness           mode     popularity    speechiness  \
count  169909.000000  169909.000000  169909.000000  169909.000000   
mean      -11.370289       0.708556      31.556610       0.094058   
std         5.666765       0.454429      21.582614       0.149937   
min       -60.000000       0.000000       0.000000       0.000000   
25%       -14.470000       0.000000      12.000000       0.034900   
50%       -10.474000       1.000000      33.000000       0.045000   
75%        -7.118000       1.000000      48.000000       0.075400   
max         3.855000       1.000000     100.000000       0.969000   

               tempo        valence           year  
count  169909.000000  169909.000000  169909.000000  
mean      116.948017       0.532095    1977.223231  
std        30.726937       0.262408      25.593168  
min         0.000000       0.000000    1921.000000  
25%        93.516000       0.322000    1957.000000  
50%       114.778000       0.544000    1978.000000  
75%       135.712000       0.749000    1999.000000  
max       244.091000       1.000000    2020.000000

In [9]:
df.dtypes

acousticness        float64
artists              object
danceability        float64
duration_ms           int64
energy              float64
explicit              int64
id                   object
instrumentalness    float64
key                   int64
liveness            float64
loudness            float64
mode                  int64
name                 object
popularity            int64
release_date         object
speechiness         float64
tempo               float64
valence             float64
year                  int64
dtype: object

## Are there missing values ?

In [29]:
df.isna().any().any()

False

# Data Restructuration

In [31]:
# remove the release_date column because of the inconsistency of its data
# sometimes we have the whole date and other times, it's just the year
# it seems like just keeping the year column is relevant
new_df = df.drop(columns=["release_date"])

In [32]:
new_df.shape

(169909, 18)

# Correlation between columns

In [39]:
# compute the pairwise correlation coefficient between all the columns but name, artists, year, id, key
new_df2 = new_df.drop(columns=["name", "artists", "year", "id", "key"])
new_df2.corr(numeric_only=False)

acousticness  danceability  duration_ms    energy  explicit  \
acousticness          1.000000     -0.265950    -0.079311 -0.750283 -0.253690   
danceability         -0.265950      1.000000    -0.134500  0.220569  0.241891   
duration_ms          -0.079311     -0.134500     1.000000  0.036396 -0.043811   
energy               -0.750283      0.220569     0.036396  1.000000  0.142677   
explicit             -0.253690      0.241891    -0.043811  0.142677  1.000000   
instrumentalness      0.335821     -0.281429     0.084814 -0.287692 -0.138292   
liveness             -0.023871     -0.105532     0.034270  0.126293  0.039272   
loudness             -0.567072      0.294170    -0.014687  0.782982  0.152695   
mode                  0.046475     -0.045306    -0.046981 -0.038355 -0.083221   
popularity           -0.593345      0.221077     0.063292  0.497488  0.214044   
speechiness          -0.056077      0.225305    -0.058449 -0.045226  0.413074   
tempo                -0.204982     -0.004872    -0.028816  0.249936  0.011484   
valence              -0.185540      0.560242    -0.198760  0.350086 -0.022327   

                  instrumentalness  liveness  loudness      mode  popularity  \
acousticness              0.335821 -0.023871 -0.567072  0.046475   -0.593345   
danceability             -0.281429 -0.105532  0.294170 -0.045306    0.221077   
duration_ms               0.084814  0.034270 -0.014687 -0.046981    0.063292   
energy                   -0.287692  0.126293  0.782982 -0.038355    0.497488   
explicit                 -0.138292  0.039272  0.152695 -0.083221    0.214044   
instrumentalness          1.000000 -0.047397 -0.417033 -0.035051   -0.299829   
liveness                 -0.047397  1.000000  0.052985  0.005393   -0.075293   
loudness                 -0.417033  0.052985  1.000000 -0.013147    0.466546   
mode                     -0.035051  0.005393 -0.013147  1.000000   -0.032854   
popularity               -0.299829 -0.075293  0.466546 -0.032854    1.000000   
speechiness              -0.115735  0.147667 -0.105796 -0.057493   -0.135707   
tempo                    -0.107570  0.008124  0.211114  0.014539    0.135047   
valence                  -0.193929 -0.000426  0.308418  0.014727    0.009327   

                  speechiness     tempo   valence  
acousticness        -0.056077 -0.204982 -0.185540  
danceability         0.225305 -0.004872  0.560242  
duration_ms         -0.058449 -0.028816 -0.198760  
energy              -0.045226  0.249936  0.350086  
explicit             0.413074  0.011484 -0.022327  
instrumentalness    -0.115735 -0.107570 -0.193929  
liveness             0.147667  0.008124 -0.000426  
loudness            -0.105796  0.211114  0.308418  
mode                -0.057493  0.014539  0.014727  
popularity          -0.135707  0.135047  0.009327  
speechiness          1.000000 -0.010070  0.056383  
tempo               -0.010070  1.000000  0.171182  
valence              0.056383  0.171182  1.000000

From this pairwise correlation, the data below have a "strong" correlation(there, we consider as 
strong correlation a correlation for which Pearson coefficient r is >= 0.5) : 
- acousticness and energy are strongly correlated with r = -0.75
- acousticness is also correlated to loudness with r = -0.56
- acousticness and popularity are correlated with r = -0.59
- danceability is correlated to valence with r = 0.56
- energy is strongly correlated to loudness with r = 0.78

From those observations, some questions pop up :
- will PCA be enough to visualize the underlying structure of those data ? In fact, 8/13 columns are pairwise correlated. And we know that PCA is based on the linear dependence of features. So, is there any linear dependence between all those data ?
- can we use acoustic to predict energy, loudness and/or popularity ?
- can we use danceability to predict valence ?
- can we use energy to predict loudness ?

See you on the next episode(which is the remainder of this project 😉) !!!